src/SubPool.sol

```diff
@@ -0,0 +1,135 @@
+pragma solidity 0.8.26;
+
+import {Auth} from "./Auth.sol";
+import {ISubPoolFactory} from "./interfaces/ISubPoolFactory.sol";
+import {SafeTransferLib} from "solady/utils/SafeTransferLib.sol";
+import {ERC20} from "solady/tokens/ERC20.sol";
+
+using SafeTransferLib for address;
+
+contract SubPool is Auth {
+    error SubPool__OnlyFactory();
+    error SubPool__CollateralTokenAlreadyInitialized();
+    error SubPool__InsufficientETH();
+    error SubPool__InvalidWithdraw();
+
+    /// @dev cannot be immutable because that allows for multiple pools per solver address.
+    address public collateralToken;
+    address public immutable COW;
+    ISubPoolFactory public immutable factory;
+
+    uint256 collateralDue;
+    uint256 cowDue;
+    uint256 ethDue;
+
+    modifier onlyFactory() {
+        if (msg.sender != address(factory)) {
+            revert SubPool__OnlyFactory();
+        }
+        _;
+    }
+
+    constructor(address _owner, address _cow) {
+        factory = ISubPoolFactory(msg.sender);
+        COW = _cow;
+
+        // rely the owner
+        _addOwner(_owner);
+    }
+
+    function initializeCollateralToken(address token) external {
+        if (collateralToken != address(0)) revert SubPool__CollateralTokenAlreadyInitialized();
+        collateralToken = token;
+    }
+
+    /// @notice Signal the intent to announce exit the solver pool.
+    function announceExit() external auth {
+        factory.announceExit();
+    }
+
+    /// @notice withdraw arbitrary tokens
+    /// @dev Can only withdraw non cow and non collateral tokens while the pool is active or in exit delay.
+    ///      When the pool's exit delay has elapsed it can withdraw any token and ether balance.
+    function withdrawTokens(address[] calldata tokens) external auth {
+        uint256 exitTimestamp = factory.exitTimestamp(address(this));
+        bool exitElapsed = exitTimestamp != 0 && block.timestamp >= exitTimestamp;
+
+        if (exitElapsed) {
+            for (uint256 i = 0; i < tokens.length;) {
+                tokens[i].safeTransfer(msg.sender, ERC20(tokens[i]).balanceOf(address(this)));
+                unchecked {
+                    ++i;
+                }
+            }
+            uint256 ethBalance = address(this).balance;
+            if (ethBalance > 0) {
+                msg.sender.safeTransferETH(address(this).balance);
+            }
+        } else {
+            for (uint256 i = 0; i < tokens.length;) {
+                address token = tokens[i];
+                if (token == COW || token == collateralToken) revert SubPool__InvalidWithdraw();
+                token.safeTransfer(msg.sender, ERC20(token).balanceOf(address(this)));
+                unchecked {
+                    ++i;
+                }
+            }
+        }
+    }
+
+    /// @notice Pull required number of tokens from the sender to push the pool
+    ///         back to the collateralization requirements.
+    function heal() external payable {
```

Comment:
Is this function needed after all changes we did?
We check if the collateral amounts are right off-chain. To "heal" the pool the owner should just transfer the right amounts. Here the due amounts are misleading because the contract can't know how much it's actually due (think about the CIP's initial deposit to target deposit).

Reply:
We are already assuming this via `canSolve` that any dues(the billed amounts) are the only dues needed to be paid to be considered collateralized?

Reply:
True, but the name `canSolve` is misleading as it's the offchain part that determines if a solver can solve already. An on-chain check for some arbitrary dues isn't necessary imho.
For example, if a pool is created with no collateral at the start, `canSolve` returns true and there is no way to tell the contract that there are dues pending.

Reply:
This is technically just to check if anything that was billed was deposited again. Without this information and condition, `canSolve` will basically just be a pool membership and pool not announced exit check. Should we rename `heal` to something else or remove it altogether?

Reply:
Yep, everything is as you say.
My preference goes towards removing it, but I'd like some feedback from @mfw78 and @fleupold to confirm this.

Reply:
It does seem that there is a bias that if functionality is questionable, it just be removed and done off-chain. In which case, same preference - let's remove `heal`. Whoever implements the off-chain component can handle it.

Reply:
removed